### PR TITLE
Document how to use high level synthesis plugins

### LINKED ISSUE #505: HLSConfig missing from API docs
### URL to the relevant documentation

Related: https://github.com/Qiskit/documentation/issues/238

### Select all that apply

- [ ] typo
- [ ] code bug
- [ ] out-of-date content
- [ ] broken link
- [X] other

### Describe the fix.

[HLSConfig](https://github.com/Qiskit/qiskit/blob/523ea748b34cc1c007eda6e9e673ab6742b304a0/qiskit/transpiler/passes/synthesis/high_level_synthesis.py#L53) should appear in the API docs.

### LINKED ISSUE #11356: Document usage of high level synthesis plugins
### What should we add?

The [API documentation for synthesis plugins](https://docs.quantum.ibm.com/api/qiskit/transpiler_synthesis_plugins) does not explain how to use high level synthesis plugins. For example, the [Using plugins](https://docs.quantum.ibm.com/api/qiskit/transpiler_synthesis_plugins#using-plugins) section only discusses unitary synthesis plugins, not high level synthesis plugins.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds missing documentation to the transpiler synthesis plugin page on how to use the high level synthesis plugin interface. There was existing documentation on how to write a high level synthesis plugin but not for how users can consume any installed plugins. This commit fills that documentation gap and also adds the HLSConfig class to the documentation as a stable defined interface. This was an oversight as we've not been treating the API interface as stable as it's the primary entrypoint for using high level synthesis plugins. Additionally, a new function high_level_synthesis_plugin_names is added in this PR to make querying the installed plugins a bit easier as to do this before you had to instantiate a plugin manager object to do this. This makes the querying interface similar to what exists for unitary synthesis.

### Details and comments

Fixes #11356
Fixes https://github.com/Qiskit/documentation/issues/505

### CODE REVIEW

File: qiskit/transpiler/passes/synthesis/plugin.py [hunk 1]
```diff
@@ -243,6 +246,36 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
 :func:`qiskit.transpiler.passes.synthesis.plugin.unitary_synthesis_plugin_names`
 function.
 
+High Level Synthesis Plugins
+----------------------------
+
+To use a high level synthesis you leverage the :class:`.HLSConfig` class to
+create an input configuration for how to synthesize multiple high level
+plugins. For example::
```
Comment (kevinsung) [inline qiskit/transpiler/passes/synthesis/plugin.py]:
```suggestion
To use a high-level synthesis plugin, you first instantiate an :class:`.HLSConfig` to
store the names of the plugins to use for various high-level objects.
For example::
```

File: qiskit/transpiler/passes/synthesis/plugin.py [hunk 2]
```diff
@@ -243,6 +246,36 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
 :func:`qiskit.transpiler.passes.synthesis.plugin.unitary_synthesis_plugin_names`
 function.
 
+High Level Synthesis Plugins
```
Comment (kevinsung) [inline qiskit/transpiler/passes/synthesis/plugin.py]:
```suggestion
High-level Synthesis Plugins
```

File: qiskit/transpiler/passes/synthesis/plugin.py [hunk 3]
```diff
@@ -243,6 +246,36 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
...
+    HLSConfig(permutation=["acg"], clifford=["layers"], linear_function=["pmh"])
+
+will create a high level synthesis configuration that will use the ``acg`` plugin
+for synthesizing :class:`.PermutationGate` objects, the ``layers`` plugin
+for synthesizing :class:`.Clifford` objects, and the ``pmh`` plugin for synthesizing
+:class:`.LinearFunction` objects. The :class:`.HLSConfig` class
+has the flexibility to specify multiple plugins in sequence which will enable
+trying multiple synthesis plugins and using the first one that successsfully
+synthesizes the object and also to pass tuples of the plugin along with options
+to adjust the behavior of the synthesis plugin.
```
Comment (kevinsung) [inline qiskit/transpiler/passes/synthesis/plugin.py]:
```suggestion
creates a high-level synthesis configuration that uses the ``acg`` plugin
for synthesizing :class:`.PermutationGate` objects, the ``layers`` plugin
for synthesizing :class:`.Clifford` objects, and the ``pmh`` plugin for synthesizing
:class:`.LinearFunction` objects.
For each high-level object, the list of given plugins are tried in sequence until one of them succeeds (in the example above, each list only contains a single plugin). In addition to specifying a plugin by its name, one can also instead pass a ``(name, options)`` tuple, where the second element of the tuple contains options for the plugin.
```

@mtreinish I suggested a significant rewording. Is it correct? Also, what is the format of the options for the plugin?

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: qiskit/transpiler/passes/synthesis/plugin.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/synthesis/plugin.py]: Yeah, this looks good.

 `options` is a dictionary where the specifics are defined by a given plugin. Implementation wise they're the kwargs dict that gets passed to the plugins `.run()` method. But, for the purposes of an `HLSConfig` it's just a dictionary with arbitrary keys and values.

File: qiskit/transpiler/passes/synthesis/plugin.py [hunk 4]
```diff
@@ -243,6 +246,33 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
...
+store the names of the plugins to use for various high-level objects.
+For example::
+
+    HLSConfig(permutation=["acg"], clifford=["layers"], linear_function=["pmh"])
+
+creates a high-level synthesis configuration that uses the ``acg`` plugin
+for synthesizing :class:`.PermutationGate` objects, the ``layers`` plugin
+for synthesizing :class:`.Clifford` objects, and the ``pmh`` plugin for synthesizing
+:class:`.LinearFunction` objects.
+For each high-level object, the list of given plugins are tried in sequence until one of them succeeds (in the example above, each list only contains a single plugin). In addition to specifying a plugin by its name, one can also instead pass a ``(name, options)`` tuple, where the second element of the tuple contains options for the plugin.
```
Comment (kevinsung) [inline qiskit/transpiler/passes/synthesis/plugin.py]:
```suggestion
For each high-level object, the list of given plugins are tried in sequence until one of them succeeds (in the example above, each list only contains a single plugin). In addition to specifying a plugin by its name, you can instead pass a ``(name, options)`` tuple, where the second element of the tuple is a dictionary containing options for the plugin.
```

File: qiskit/transpiler/passes/synthesis/plugin.py [hunk 5]
```diff
@@ -243,6 +246,33 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
...
 function.
 
+High-level Synthesis Plugins
+----------------------------
+
+To use a high-level synthesis plugin, you first instantiate an :class:`.HLSConfig` to
+store the names of the plugins to use for various high-level objects.
+For example::
+
+    HLSConfig(permutation=["acg"], clifford=["layers"], linear_function=["pmh"])
```
Comment (kevinsung) [inline qiskit/transpiler/passes/synthesis/plugin.py:258]: Can you comment on the possible keyword arguments here? e.g. how does the user know that `permutation`, `clifford`, and `linear_function` are possibilities?

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: qiskit/transpiler/passes/synthesis/plugin.py (hunk 5 above)
Comment (jakelishman) [reply qiskit/transpiler/passes/synthesis/plugin.py:258]: I've pushed up a8244435e2 to expand on this - is it understandable for you?

**Comment by jakelishman** [comment]
I'll take over the last few changes needed for this on Matt's behalf, since he's on holiday now and I'd like to get the documentation whole plugged sooner rather than later.  Let me know if there's anything more that needs changing, and I can do it.

**Comment by kevinsung** [comment]
Thanks, it's pretty clear now. Now the documentation here is clearer than in the docstring of HLSConfig. How about we add a link in that docstring to this documentation?

**Review by kevinsung — DISMISSED** [review]
LGTM with just one comment about improving the docstring of HLSConfig itself

**Comment by jakelishman** [comment]
Sure, I've tried to make `HLSConfig` a little clearer about the allowed forms, and put in a cross-link to the overview section in 5cc74e18aa.

**Review by kevinsung — APPROVED** [review]
Thanks!

**Review by jakelishman — APPROVED** [review]
(proxying approval)

**Pull request merged**